hazelcast/src/test/java/com/hazelcast/instance/HazelcastInstanceFactoryTest.java

```diff
@@ -43,62 +39,80 @@
+                @Override
+                public void run() throws Exception {
+                    assertEquals(3, instance1.getCluster().getMembers().size());
+                    assertEquals(3, instance2.getCluster().getMembers().size());
+                    assertEquals(3, instance3.getCluster().getMembers().size());
+                }
+            });
+        } finally {
+            instanceFactory.terminateAll();
+        }
+    }
 
+    @Test
+    public void testTestHazelcastInstanceFactory_withTwoFactories() {
+        TestHazelcastInstanceFactory instanceFactory1 = new TestHazelcastInstanceFactory();
         TestHazelcastInstanceFactory instanceFactory2 = new TestHazelcastInstanceFactory();
-        final HazelcastInstance instance1_2 = instanceFactory2.newHazelcastInstance();
-        final HazelcastInstance instance2_2 = instanceFactory2.newHazelcastInstance();
-
-
-        assertTrueEventually(new AssertTask() {
-            @Override
-            public void run() throws Exception {
-                assertEquals(2, instance1_2.getCluster().getMembers().size());
-                assertEquals(2, instance2_2.getCluster().getMembers().size());
-            }
-        });
-
-        assertTrueEventually(new AssertTask() {
-            @Override
-            public void run() throws Exception {
-                assertEquals(3, instance1.getCluster().getMembers().size());
-                assertEquals(3, instance2.getCluster().getMembers().size());
-                assertEquals(3, instance3.getCluster().getMembers().size());
-            }
-        });
-
-        instanceFactory.terminateAll();
-        instanceFactory2.terminateAll();
-
+        try {
+            final HazelcastInstance instance11 = instanceFactory1.newHazelcastInstance();
+            final HazelcastInstance instance12 = instanceFactory1.newHazelcastInstance();
+            final HazelcastInstance instance13 = instanceFactory1.newHazelcastInstance();
+
+            final HazelcastInstance instance21 = instanceFactory2.newHazelcastInstance();
+            final HazelcastInstance instance22 = instanceFactory2.newHazelcastInstance();
+
+            assertTrueEventually(new AssertTask() {
+                @Override
+                public void run() throws Exception {
+                    assertEquals(2, instance21.getCluster().getMembers().size());
+                    assertEquals(2, instance22.getCluster().getMembers().size());
+                }
+            });
+            assertTrueEventually(new AssertTask() {
+                @Override
+                public void run() throws Exception {
+                    assertEquals(3, instance11.getCluster().getMembers().size());
+                    assertEquals(3, instance12.getCluster().getMembers().size());
+                    assertEquals(3, instance13.getCluster().getMembers().size());
+                }
+            });
+        } finally {
+            instanceFactory1.terminateAll();
```

Comment:
I moved the `TestHazelcastInstanceFactory` shutdowns to a finally block to ensure they're called even if an exception is thrown in the test. Since they're not used in all tests, I didn't pull them out as class fields.